lib/saml/auth_fail_handler.rb

```diff
@@ -17,15 +19,17 @@ def initialize(saml_response, user, session)
       @context = nil
     end
 
-    def known_error?
-      known_errors = [
-        :clicked_deny?,
-        :auth_too_late?,
-        :auth_too_early?
-      ]
+    def error
+      KNOWN_ERRORS.each do |known_error|
+        return known_error if send("#{known_error}?")
+      end
 
-      known_errors.each do |known_error|
-        break if send(known_error)
+      'unknown'
+    end
+
+    def known_error?
+      KNOWN_ERRORS.each do |known_error|
+        break if send("#{known_error}?")
```

Comment:
and then again, we're iterating over all the known errors, where I think we could just do a hash based lookup to determine what the error is without having to iterate.